File: android/src/main/kotlin/com/chyiiiiiiiiiiiiii/zendesk_messaging/ZendeskMessaging.kt
import android.content.Intent
import com.chyiiiiiiiiiiiiii.zendesk_messaging.ZendeskMessagingPlugin
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import zendesk.android.Zendesk
import zendesk.android.ZendeskUser
import zendesk.android.events.ZendeskEvent
import zendesk.android.events.ZendeskEventListener
import zendesk.messaging.android.DefaultMessagingFactory


class ZendeskMessaging(private val plugin: ZendeskMessagingPlugin, private val channel: MethodChannel) {
    companion object {
        const val tag = "[ZendeskMessaging]"

        // Method channel callback keys
        const val initializeSuccess: String = "initialize_success"
        const val initializeFailure: String = "initialize_failure"
        const val loginSuccess: String = "login_success"
        const val loginFailure: String = "login_failure"
        const val logoutSuccess: String = "logout_success"
        const val logoutFailure: String = "logout_failure"

        // add start
        const val unreadMessageCountChanged: String = "unread_message_count_changed"
        const val authenticationFailed: String = "authentication_failed"
        const val unknownEvent: String = "unknown_event"
        // add end
    }

    // add start
    // To create and use the event listener:
    private val zendeskEventListener: ZendeskEventListener = ZendeskEventListener { zendeskEvent ->
        when (zendeskEvent) {
            is ZendeskEvent.UnreadMessageCountChanged -> {
                val unreadCount = getUnreadMessageCount()
                channel.invokeMethod(unreadMessageCountChanged, mapOf("unreadCount" to unreadCount))
            }

            is ZendeskEvent.AuthenticationFailed -> {
                channel.invokeMethod(authenticationFailed, null)
            }

            is ZendeskEvent.FieldValidationFailed -> {
                channel.invokeMethod(unknownEvent, null)
            }

            else -> {
                channel.invokeMethod(unknownEvent, null)
            }
        }
    }
    // add end


    fun initialize(channelKey: String) {
        println("$tag - Channel Key - $channelKey")
        Zendesk.initialize(
                plugin.activity!!,
                channelKey,
                successCallback = { value ->
                    plugin.isInitialized = true;
                    println("$tag - initialize success - $value")
                    channel.invokeMethod(initializeSuccess, null)
                    // add start
                    Zendesk.instance.addEventListener(zendeskEventListener)
                    // add end
                },
                failureCallback = { error ->
                    plugin.isInitialized = false
                    println("$tag - initialize failure - $error")
                    channel.invokeMethod(initializeFailure, mapOf("error" to error.message))
                },
                messagingFactory = DefaultMessagingFactory()
        )
    }

    fun invalidate() {
        Zendesk.invalidate()
        plugin.isInitialized = false;
        // add start
        Zendesk.instance.removeEventListener(zendeskEventListener)
        // add end
        println("$tag - invalidated")
    }

    fun show() {
        Zendesk.instance.messaging.showMessaging(plugin.activity!!, Intent.FLAG_ACTIVITY_NEW_TASK)
        println("$tag - show")
    }

    fun getUnreadMessageCount(): Int {
        return try {
            Zendesk.instance.messaging.getUnreadMessageCount()
        } catch (error: Throwable) {
            0
        }
    }

    fun setConversationTags(tags: List<String>) {
        Zendesk.instance.messaging.setConversationTags(tags)
    }

    fun clearConversationTags() {
        Zendesk.instance.messaging.clearConversationTags()
    }

    fun loginUser(jwt: String) {
        Zendesk.instance.loginUser(
                jwt,
                { value: ZendeskUser? ->
                    plugin.isLoggedIn = true;
                    value?.let {
                        channel.invokeMethod(loginSuccess, mapOf("id" to it.id, "externalId" to it.externalId))
                    } ?: run {
                        channel.invokeMethod(loginSuccess, mapOf("id" to null, "externalId" to null))
                    }
                },
                { error: Throwable? ->
                    println("$tag - Login failure : ${error?.message}")
                    println(error)
                    channel.invokeMethod(loginFailure, mapOf("error" to error?.message))
                })
    }

    fun logoutUser() {
        GlobalScope.launch(Dispatchers.Main) {
            try {
                Zendesk.instance.logoutUser(successCallback = {
                    plugin.isLoggedIn = false;
                    channel.invokeMethod(logoutSuccess, null)
                }, failureCallback = {
                    channel.invokeMethod(logoutFailure, null)
                });
            } catch (error: Throwable) {
                println("$tag - Logout failure : ${error.message}")
                channel.invokeMethod(logoutFailure, mapOf("error" to error.message))
            }
        }
    }

    fun setConversationFields(fields: Map<String, String>) {
        Zendesk.instance.messaging.setConversationFields(fields)
    }

    fun clearConversationFields() {
        Zendesk.instance.messaging.clearConversationFields()
    }
}

File: ios/Classes/ZendeskMessaging.swift
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK

public class ZendeskMessaging: NSObject {
    private static var initializeSuccess: String = "initialize_success"
    private static var initializeFailure: String = "initialize_failure"
    private static var loginSuccess: String = "login_success"
    private static var loginFailure: String = "login_failure"
    private static var logoutSuccess: String = "logout_success"
    private static var logoutFailure: String = "logout_failure"
    // add start
    private static var unreadMessageCountChanged: String = "unread_message_count_changed"
    private static var authenticationFailed: String = "authentication_failed"
    private static var unknownEvent: String = "unknown_event"
    // add end
    
    let TAG = "[ZendeskMessaging]"
    
    private var zendeskPlugin: SwiftZendeskMessagingPlugin? = nil
    private var channel: FlutterMethodChannel? = nil

    init(flutterPlugin: SwiftZendeskMessagingPlugin, channel: FlutterMethodChannel) {
        self.zendeskPlugin = flutterPlugin
        self.channel = channel
    }
    
    func initialize(channelKey: String) {
        print("\(self.TAG) - Channel Key - \(channelKey)\n")
        Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { result in
            DispatchQueue.main.async {
                if case let .failure(error) = result {
                    self.zendeskPlugin?.isInitialized = false
                    print("\(self.TAG) - initialize failure - \(error.localizedDescription)\n")
                    self.channel?.invokeMethod(ZendeskMessaging.initializeFailure, arguments: ["error": error.localizedDescription])
                } else {
                    self.zendeskPlugin?.isInitialized = true
                    print("\(self.TAG) - initialize success")
                    self.channel?.invokeMethod(ZendeskMessaging.initializeSuccess, arguments: [:])
                    
                    // add start
                    Zendesk.instance?.addEventObserver(self) { event in
                        switch event {
                        case .unreadMessageCountChanged(let unreadCount):
                            self.channel?.invokeMethod(ZendeskMessaging.unreadMessageCountChanged, arguments: ["unreadCount": unreadCount])
                        case .authenticationFailed(let error as NSError):
                            print("Authentication error received: \(error)")
                            print("Domain: \(error.domain)")
                            print("Error code: \(error.code)")
                            print("Localized Description: \(error.localizedDescription)")
                            self.channel?.invokeMethod(ZendeskMessaging.authenticationFailed, arguments: ["error": error.localizedDescription])
                        @unknown default:
                            self.channel?.invokeMethod(ZendeskMessaging.unknownEvent, arguments: ["event": event])
                            break
                        }
                    }
                    // add end
                }
            }
        }
    }

    func invalidate() {
        Zendesk.invalidate()
       self.zendeskPlugin?.isInitialized = false
       // add start
       Zendesk.instance?.removeEventObserver(self)
       // add end
       print("\(self.TAG) - invalidate")
    }
    
    func show(rootViewController: UIViewController?) {
        guard let messagingViewController = Zendesk.instance?.messaging?.messagingViewController() as? UIViewController else {
            print("\(self.TAG) - Unable to create Zendesk messaging view controller")
            return
        }
        guard let rootViewController = rootViewController else {
            print("\(self.TAG) - Root view controller is nil")
            return
        }
        
        messagingViewController.modalPresentationStyle = .fullScreen

        // Check if rootViewController is already presenting another view controller
        if let presentedVC = rootViewController.presentedViewController {
            // Check if the presentedVC is the same instance as messagingViewController
            if presentedVC === messagingViewController {
                // If the same instance, do nothing or update it as necessary
                print("\(self.TAG) - Zendesk messaging view controller is already presented")
            } else {
                // Dismiss current and present new, or just present new
                presentedVC.dismiss(animated: true) {
                    rootViewController.present(messagingViewController, animated: true, completion: nil)
                }
            }
        } else {
            // No view controller is being presented, present the new one
            rootViewController.present(messagingViewController, animated: true, completion: nil)
        }

        print("\(self.TAG) - show")
    }
    
    func sendPageViewEvent(pageTitle: String, url: String, completionHandler: @escaping (Result<Void, Error>) -> Void) {
        let pageView = PageView(pageTitle: pageTitle, url: url)
        Zendesk.instance?.sendPageViewEvent(pageView,completionHandler: completionHandler)
    }

    func setConversationTags(tags: [String]) {
        Zendesk.instance?.messaging?.setConversationTags(tags)
    }

    func clearConversationTags() {
        Zendesk.instance?.messaging?.clearConversationTags()
    }
    
    func loginUser(jwt: String) {
        Zendesk.instance?.loginUser(with: jwt) { result in
            DispatchQueue.main.async {
                switch result {
                case .success(let user):
                    self.zendeskPlugin?.isLoggedIn = true
                    self.channel?.invokeMethod(ZendeskMessaging.loginSuccess, arguments: ["id": user.id, "externalId": user.externalId])
                    break
                case .failure(let error):
                    print("\(self.TAG) - login failure - \(error.localizedDescription)\n")
                    self.channel?.invokeMethod(ZendeskMessaging.loginFailure, arguments: ["error": nil])
                    break
                }
            }
        }
    }
    
    func logoutUser() {
        Zendesk.instance?.logoutUser { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    self.zendeskPlugin?.isLoggedIn = false
                    self.channel?.invokeMethod(ZendeskMessaging.logoutSuccess, arguments: [])
                    break
                case .failure(let error):
                    print("\(self.TAG) - logout failure - \(error.localizedDescription)\n")
                    self.channel?.invokeMethod(ZendeskMessaging.logoutFailure, arguments: ["error": nil])
                    break
                }
            }
        }
    }
    
    func getUnreadMessageCount() -> Int {
        let count = Zendesk.instance?.messaging?.getUnreadMessageCount()
        return count ?? 0
    }

    func setConversationFields(fields: [String: String]) {
        Zendesk.instance?.messaging?.setConversationFields(fields)
    }

    func clearConversationFields() {
        Zendesk.instance?.messaging?.clearConversationFields()
    }
}
